pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,607 @@
+	Region          string `hcl:"region" json:"region"`
+	KeyPrefix       string `hcl:"key_prefix" json:"key_prefix"`
+}
+
+// New returns an instantiated plugin
+func New() *Plugin {
+	return newPlugin(newKMSClient)
+}
+
+func newPlugin(newClient func(ctx context.Context, config *Config) (kmsClient, error)) *Plugin {
+	return &Plugin{
+		entries: make(map[string]keyEntry),
+		hooks: pluginHooks{
+			newClient: newClient,
+		},
+		scheduleDelete: make(chan string, 120),
+	}
+}
+
+// SetLogger sets a logger
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure sets up the plugin
+func (p *Plugin) Configure(ctx context.Context, req *plugin.ConfigureRequest) (*plugin.ConfigureResponse, error) {
+	// cancels previous schedule delete task in case of re configure
+	if p.cancelScheduleDelete != nil {
+		p.cancelScheduleDelete()
+	}
+
+	config, err := validateConfig(req.Configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.keyPrefix = config.KeyPrefix
+	p.kmsClient, err = p.hooks.newClient(ctx, config)
+	if err != nil {
+		return nil, newErrorf(codes.Internal, "failed to create KMS client: %v", err)
+	}
+
+	p.log.Debug("Fetching key aliases from KMS")
+	err = p.buildCache(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, p.cancelScheduleDelete = context.WithCancel(context.Background())
+	go p.scheduleDeleteTask(ctx)
+
+	return &plugin.ConfigureResponse{}, nil
+}
+
+// GenerateKey creates a key in KMS. If a key already exists in the local storage, it is updated.
+func (p *Plugin) GenerateKey(ctx context.Context, req *keymanager.GenerateKeyRequest) (*keymanager.GenerateKeyResponse, error) {
+	if req.KeyId == "" {
+		return nil, newError(codes.InvalidArgument, "key id is required")
+	}
+	if req.KeyType == keymanager.KeyType_UNSPECIFIED_KEY_TYPE {
+		return nil, newError(codes.InvalidArgument, "key type is required")
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	spireKeyID := req.KeyId
+	newEntry, err := p.createKey(ctx, spireKeyID, req.KeyType)
+	if err != nil {
+		return nil, err
+	}
+
+	p.log.Debug("Key created", keyIDTag, newEntry.KMSKeyID)
+	oldEntry, hasOldEntry := p.entries[spireKeyID]
+
+	if !hasOldEntry {
+		//create alias
+		_, err = p.kmsClient.CreateAlias(ctx, &kms.CreateAliasInput{
+			AliasName:   aws.String(newEntry.Alias),
+			TargetKeyId: &newEntry.KMSKeyID,
+		})
+		if err != nil {
+			return nil, newErrorf(codes.Internal, "failed to create alias: %v", err)
+		}
+		p.log.Debug("Alias created", aliasTag, newEntry.Alias)
+	} else {
+		//update alias
+		_, err = p.kmsClient.UpdateAlias(ctx, &kms.UpdateAliasInput{
+			AliasName:   aws.String(newEntry.Alias),
+			TargetKeyId: &newEntry.KMSKeyID,
+		})
+		if err != nil {
+			return nil, newErrorf(codes.Internal, "failed to update alias: %v", err)
+		}
+		p.log.Debug("Alias updated", aliasTag, newEntry.Alias)
+
+		p.scheduleDelete <- oldEntry.KMSKeyID
+	}
+
+	err = p.setEntry(spireKeyID, newEntry)
```

Comment:
This function can fail, at which point we've already created/updated the alias (and scheduled for deletion). I wonder if we should split up this function up into validation and setting and then do the validation of this entry before it is ever returned from `createKey`....

If we do that then by the time we've created/updated the alias, we just need to set the entry, which can't fail.